src/util/simplify_expr.cpp

```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
     return false;
   }
 
-  // eliminate casts to _Bool
-  if(expr_type.id()==ID_c_bool &&
-     op_type.id()!=ID_bool)
+  // circular casts through types shorter than `int`
+  const auto check_circular_cast=[&](const typet &type)
   {
-    // casts from boolean to a signed int and back:
-    // (boolean)(int)boolean -> boolean
-    if(expr.op0().id()==ID_typecast && op_type.id()==ID_signedbv)
+    if(expr_type==type)
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      const auto &typecast=expr_dynamic_cast<typecast_exprt>(expr.op0());
+      if(typecast.op().type()==type)
       {
         expr=typecast.op0();
         return false;
       }
     }
+    return true;
+  };
+  if(op_type==signedbv_typet(32) &&
+     expr.op0().id()==ID_typecast &&
+     !(check_circular_cast(c_bool_typet(8)) &&
+       check_circular_cast(signedbv_typet(8)) &&
```

Comment:
Why not `unsignedbv_typet(8)`?

Reply:
These are supposed to correspond to the shorter-than-int types in java_types.h. However, `util` shouldn't depend on other modules so I don't know if it's OK to include that header here.

Reply:
I see the logic (and I don't think you should include the header0, but would there be anything wrong from a C perspective of disregarding a cast from an unsigned 16 -> unsigned 8 -> unsigned 16 for example? Feel free to say out of scope :)